### ISSUE TITLE
Add new data type "array" ?

### ISSUE THREAD

**Issue by eostis** [semi-technologies/weaviate#1611]
My goal here is to find a way to filter and aggregate on arrays of simple types, like arrays of strings or arrays of floats.

As I understand, one has to use [cross reference data types ](https://www.semi.technology/developers/weaviate/current/data-schema/datatypes.html#datatype-cross-reference) to be able to define arrays.

But it is not clear that [aggregations](https://www.semi.technology/developers/weaviate/current/graphql-references/aggregate.html) can work on cross reference, while [filters](https://www.semi.technology/developers/weaviate/current/graphql-references/filters.html#beacon-reference-filters) can.

**Comment by laura-ham** [comment]
@eostis Could you give an example? I don't fully understand the question.

**Comment by eostis** [comment]
Indeed.

On WooCommerce, a product can have multi-value attributes. For instance, it can be sold with colors ['blue', 'red', 'yellow'].
This product must be indexed with its 3 colors, and a facet on colors should appear in the search UI to displays available colors, with their respective counts. The colors can then be selected and used as a filter.

My question was:
1) Can we create a property 'color' with a data type 'array of strings' ?
2) Can we create a facet on the property color with the aggregation query?

If it is not possible, one could create a class 'Color', and use this class has a beacon on the 'Product' class.  
But:
3) Can we filter 'Products' (it seems so) on the class 'Color'?
4) Can we create a 'Product' facet on the beacon 'Color' with the aggregation query?

A live Woocommerce example with multi-value tags and categories: https://demo-woocommerce-flatsome-kinsta.wpsolr.com/shop/

**Comment by etiennedi** [comment]
From a technical perspective an aggregation on cross-refs would be considerably more expensive than an array data type, so it'd be in favor of that. Especially if we limit it to arrays of primitives (e.g. `string[]`, `int[]`, `number[]`, `text[]`) it should not be too complex to implement.

@eostis can you give an example of an array would be aggregated? For example on a `count` would we just count each occurrence as if it was it's own object? I.e. `obj1.colors: red, blue` and `obj2.colors: red` would be aggregated to `red: 2`, `blue: 1`, etc? Or are you looking for a different type of aggregation?

**Comment by eostis** [comment]
@etiennedi 

> I'd be in favor of arrays. Especially if we limit it to arrays of primitives

Indeed, an array of primitive types is good enough for my use cases. 

> would be aggregated to red: 2, blue: 1

Your aggregation example is correct.


Also, is it possible:
- to build a slider range facet on numerics (the min and max should be returned by the aggregation) ?
- to get a range facet on numeric: 0-10: 2, 11-*: 89

**Comment by etiennedi** [comment]
> Also, is it possible:

I'd carefully say that anything is possible, but it's always a matter of prioritization among limited resources we have. Given that aggregations are more of a "secondary" feature of Weaviate, this might take some time until we get to it. What I mean is: Users come because of the vector search and having the ability to aggregate is a convenience feature. Worst case they'd have to use a different db for the aggregations, whereas for the vector search there aren't any / hardly any alternatives.

So, additional aggregation functions might be something that we design, but rely on an open source contributor for the actual implementation at the moment.

tl;dr: I'm all for improving aggregations, but it's a bit hard to prioritize at the moment. A well made pull-request would certaintly be appreciated and accepted.

**Comment by eostis** [comment]
This makes perfect sense.

One still can use Weaviate for instant search, Q&A, or similarities, while letting Lucene-based solutions take care of the faceted search.

**Comment by bobvanluijt** [comment]
As an additional requirement, if the schema settings for an array property have (or the default setting is) `"moduleConfig": { "text2vec-*": { "vectorizePropertyName": true } }` a `string[]` or `text[]` should be used in the vectorization process.

e.g.

```
name: Jane
Job: Ice Cream Parlor Owner
```

and

```
someTextArray: ["Jane", "Ice Cream Parlor Owner"]
```

would lead to the same vector (if the property names aren’t vectorized)

**Comment by eostis** [comment]
Would it be possible to consider all data types as array by default?

This is the behavior of Elasticsearch/Solr: a simple type is declared for a field, but this field will accept simple data and array data as well.

**Comment by eostis** [comment]
For the sake of simplicity and homogeneity, it would also be great to have an array types for all simple types.

According to https://www.semi.technology/developers/weaviate/current/data-schema/datatypes.html, the following types have no array types:
- boolean
- date
- geoCoordinates
- phoneNumber
-  blob
